torchvision/datasets/utils.py

```diff
@@ -189,3 +192,37 @@ def _save_response_content(response, destination, chunk_size=32768):
                 progress += len(chunk)
                 pbar.update(progress - pbar.n)
         pbar.close()
+
+
+# thread-safe/multiprocessing-safe (unlike a Python ZipFile instance)
+class ZipLookup(object):
+    def __init__(self, filename):
+        self.root_zip_filename = filename
+        self.root_zip_lookup = {}
+
+        with zipfile.ZipFile(filename, "r") as root_zip:
+            for info in root_zip.infolist():
```

Comment:
What would be the slowdown if we completely remove creating this index beforehand, and instead in `__getitem__` we query for `root_zip.getinfo()`?

Reply:
It's significant, because on the opening of a ZipFile, it will read in the entire central index that is at the end of the ZIP-file (and parse it). For the CelebA dataset, that takes ~2 seconds on my Windows workstation.

Reply:
I see, indeed it looks like `getinfo` [reads the whole index first](https://github.com/python/cpython/blob/master/Lib/zipfile.py#L1321-L1366). I need to think about it a bit more on how to seamlessly support files without having to modify each one of the datasets with custom code.

Reply:
In commit https://github.com/pytorch/vision/pull/950/commits/3f10d56df21121512f43f9b87870a26d8ad2f205 I added a get_path_or_fp function which unifies the zip/non-zip reading - though a small exception was needed to make COCO file paths match.

Note: this is a proposal, I haven't tested it yet. What do you think?